base/applications/screensavers/3dtext/3dtext.c

```diff
@@ -360,13 +369,20 @@ ScreenSaverProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
         case WM_PAINT:
             DrawGLScene();
             SwapBuffers(hDC);
+            // Don't update here. Wait for timer.
+            ValidateRect(hWnd, NULL);
```

Comment:
Because this screensaver never actually called 'BeginPaint/EndPaint', we never marked the `WM_PAINT` as done, so probably ReactOS was spamming `WM_PAINT`s ?

It would be interesting to investigate the behavior on windows:
Does windows automatically mark a window as `done` when `WM_PAINT` returned without calling `BeginPaint/EndPaint`, or does the one calling the ScreenSaverProc do that?

Reply:
@learn-more, I can wrap this code within ```BeginPaint/EndPaint``` if you think that it is needed. It was never there before. Do you think that it is needed? Does anyone think that it is needed? Thanks.

Reply:
What @learn-more was asking is for someone to investigate e.g. what would already happen, without all this timing code but just by wrapping said code with the Begin/EndPrint alone, and see what happens (on both ReactOS and Windows).

Reply:
I am assuming that it does not have a near 100% processor usage on Windows.

From this assumption, I am speculating on why that would not be the case. I am observing a potential bug (the WM_PAINT) is never marked as 'done' in our code.
* Because I do not know if a screensaver is supposed to do this, it could be that the process that loaded the screensaver should be responsible for this (`the one calling the ScreenSaverProc`).
* Another option is that Windows detects this mis-behavior (not clearing the dirty flag after a WM_PAINT) and it either lowers the priority for the next drawing call, it it automatically marks the window as 'no longer dirty'.

Reply:
@learn-more, Your assumption is correct. After These changes Windows shows near 0% CPU usage on W2KSP2. Please review the video in CORE-17866 for proof. The video shows both the old and the new in sequential tests with the % CPU usage showing in the Task Manager's Performance tab. Thanks.

Reply:
Sorry that I did not make this clearer. When the WEBM shows 3dtextOld.scr selected, then it is running the unpatched version. When it shows 3dtextNew.scr selected it is running a version patched with this PR's changes.

![W2KSP2_3dtext_scr_UnPatched](https://user-images.githubusercontent.com/32620577/143726850-ade9e59a-c617-415b-a007-bd0ea3a7bd6e.png)

![W2KSP2_3dtext_scr_Patched](https://user-images.githubusercontent.com/32620577/143726856-e32722a6-6473-4d41-91bf-990b48c0a9e8.png)

Reply:
@HBelusca, So the answer to your question "But does our original screensaver (without any fixes whatsoever) uses ~ 100% CPU on windows itself?" is definitely "Yes". The proof is in the WEBM. Thanks.

Reply:
ok, so our screensaver is clearly bugged, which would make sense if it tries to refresh at top speed.

Thanks.

Reply:
@Doug-Lyons thanks for your investigations!
What I still wonder however, based on the suggestion by @learn-more , is whether the original screensaver, with the addition of the following minimal patch (without all the WM_TIMER thing, i.e. this is NOT the current PR!) would also reduce the CPU usage on both ReactOS and Windows:
```c
        case WM_PAINT:
        {
            PAINTSTRUCT ps;
            BeginPaint(hWnd, &ps);
            DrawGLScene();
            SwapBuffers(hDC);
            EndPaint(hWnd, &ps);
            break;
```

EDIT: Heh! Other people have also encountered this problem! See for example:
https://stackoverflow.com/questions/2842319/swapbuffers-causes-redraw
https://docs.microsoft.com/en-us/windows/win32/opengl/the-program-ported-to-win32
https://www.khronos.org/opengl/wiki/Platform_specifics:_Windows

Reply:
I tried to attach a video, but received a message that the file type was not accepted.
Here is a screen shot. There is no animation and no movement of the sceensaver either in the small display or full screen.
I will try and attach the video in the CORE Jira Report.

![W2KSP2_3dtextHB_scr](https://user-images.githubusercontent.com/32620577/143784211-31122d93-176c-43a9-b4ab-38ac9a1ee841.png)

Reply:
Thanks @Doug-Lyons . So the conclusion seems, that we either need to fix the screensaver the way you did it, or using an alternative like in the links I posted above.